IdentityCore/src/network/request/MSIDAADAuthorityMetadataRequest.m

```diff
@@ -0,0 +1,105 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDAADAuthorityMetadataRequest.h"
+#import "MSIDAADResponseSerializer.h"
+#import "MSIDAADRequestConfigurator.h"
+#import "MSIDAADNetworkConfiguration.h"
+
+@interface MSIDAADAuthorityMetadataResponseSerializer : MSIDAADResponseSerializer
+@end
+
+@implementation MSIDAADAuthorityMetadataResponseSerializer
+
+- (id)responseObjectForResponse:(NSHTTPURLResponse *)httpResponse
+                           data:(NSData *)data
+                        context:(id <MSIDRequestContext>)context
+                          error:(NSError **)error
+{
+    NSError *jsonError;
+    NSMutableDictionary *jsonObject = [[super responseObjectForResponse:httpResponse data:data context:context error:&jsonError] mutableCopy];
+
+    if (jsonError)
+    {
+        if (error) *error = jsonError;
+        return nil;
+    }
+
+    __auto_type reponse = [MSIDAADAuthorityMetadataResponse new];
+    
+    NSString *oauthError = jsonObject[@"error"];
+    if (![NSString msidIsStringNilOrBlank:oauthError])
+    {
+        if (error) {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorDeveloperAuthorityValidation,
+                                     jsonObject[@"error_description"],
+                                     oauthError,
+                                     nil, nil, context.correlationId, nil);
+        }
+        
+        return nil;
+    }
+
+    reponse.metadata = jsonObject[@"metadata"];
+    reponse.openIdConfigurationEndpoint = [NSURL URLWithString:jsonObject[@"tenant_discovery_endpoint"]];
+
+    return reponse;
+}
+
+@end
+
+@implementation MSIDAADAuthorityMetadataResponse
+@end
+
+@implementation MSIDAADAuthorityMetadataRequest
+
+- (instancetype)initWithEndpoint:(NSURL *)endpoint
+                       authority:(NSURL *)authority
+{
+    self = [super init];
+    if (self)
+    {
+        NSParameterAssert(endpoint);
+        NSParameterAssert(authority);
+        
+        NSMutableDictionary *parameters = [NSMutableDictionary new];
+        parameters[@"api-version"] = MSIDAADNetworkConfiguration.defaultConfiguration.aadAuthorityDiscoveryApiVersion;
+        __auto_type authorizationEndpoint = [MSIDAADNetworkConfiguration.defaultConfiguration.endpointProvider oauth2AuthorizeEndpointWithUrl:authority].absoluteString;
+        parameters[@"authorization_endpoint"] = authorizationEndpoint;
+        _parameters = parameters;
```

Comment:
yes, MSIDUrlRequestSerializer does serialization of parameters using NSURLComponents.

Reply:
NSURLComponents doesn't necessarily percent encode the url. @unpluggedk Saw some issues because of that in ADFS. Just wanted to verify, because previously authorization_endpoint was percent encoded.